<h1><b>Domaci zadatak 2 iz predmeta "Genomska informatika"</b1></h1>
<h2>Student: Žarko Babić 3109/23</h2>

<h3>Installing the pysam library:</h3>

In [1]:
pip install pysam

Defaulting to user installation because normal site-packages is not writeable
Note: you may need to restart the kernel to use updated packages.


<h3>Taking the first read from the AlignmentFile:</h3>

In [2]:
import pysam
bamFile = pysam.AlignmentFile("merged-tumor.bam")
firstRead = None

for read in bamFile:
    firstRead = read
    break
bamFile.close()

<h3>Inspecting the fields of the AlignedSegment object made from the first read:</h3>

<i><b>QNAME:</b></i>

In [3]:
print(firstRead.query_name)

C0HVYACXX120402:7:1207:5722:57044


<i><b>FLAG:</b></i>

In [4]:
print(firstRead.flag)

1187


<i><b>RNAME:</b></i>

In [5]:
print(firstRead.reference_id)

20


<i><b>POS:</b></i>

In [6]:
print(firstRead.reference_start)

9483248


<i><b>MAPQ:</b></i>

In [7]:
print(firstRead.mapping_quality)

27


<i><b>CIGAR:</b></i>

In [8]:
print(firstRead.cigarstring)

76M


<i><b>MRNM: </b></i>(the reference id of the mate/next read)

In [9]:
print(firstRead.next_reference_id)

20


<i><b>MPOS:</b></i> (the position of the mate/next read)

In [10]:
print(firstRead.next_reference_start)

9483381


<i><b>ISIZE:</b></i>

In [11]:
print(firstRead.template_length)

209


<i><b>SEQ:</b></i>

In [12]:
print(firstRead.seq)

TTTTCAAACAGTATCTATGCCTGCCAAATGTGAACATATAAAAAAAAACCAGAATGTGCCATTCTGATTTAAACTG


<i><b>QUAL:</b></i>

In [13]:
print(firstRead.qual)

==<>@?@@>@D?>@C?>8JA5?>CC>?@???BB;H-:4A?D=DB8BDE?GBJCD@B8??E<A>CDJB@B>AA@@@C


<i><b>Decoded qualities (firstRead.qual(qualities in ASCII) - 33):</b></i>

In [14]:
print(firstRead.query_qualities)

array('B', [28, 28, 27, 29, 31, 30, 31, 31, 29, 31, 35, 30, 29, 31, 34, 30, 29, 23, 41, 32, 20, 30, 29, 34, 34, 29, 30, 31, 30, 30, 30, 33, 33, 26, 39, 12, 25, 19, 32, 30, 35, 28, 35, 33, 23, 33, 35, 36, 30, 38, 33, 41, 34, 35, 31, 33, 23, 30, 30, 36, 27, 32, 29, 34, 35, 41, 33, 31, 33, 29, 32, 32, 31, 31, 31, 34])


<i><b>TAG:</b></i>

In [15]:
print(firstRead.get_tags())

[('XA', 'GL000217.1,-110754,76M,1;'), ('BD', 'NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN'), ('MD', '76'), ('RG', '1'), ('BI', 'NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN'), ('NM', 0), ('MQ', 27), ('AS', 76), ('XS', 71)]


<h3>Inspecting the flag field:</h3>

<b>Flag field of the first read:</b>

In [16]:
print(firstRead.flag)

1187


<b>Summary of the flag 1187:</b>
    <li>read paired (0x1)</li>
    <li>read mapped in proper pair (0x2)</li>
    <li>mate reverse strand (0x20)</li>
    <li>second in pair (0x80)</li>
    <li>read is PCR or optical duplicate (0x400)</li>

<h3>Checking out the flag for some reads:</h3>

In [17]:
counter = 0
mask_read_paired = 0x1
mask_read_mapped_in_proper_pair = 0x2
mask_read_unmapped = 0x4
mask_mate_unmapped = 0x8
mask_read_reverse_strand = 0x10
mask_mate_reverse_strand = 0x20
mask_first_in_pair = 0x40
mask_second_in_pair = 0x80
mask_not_primary_alignment = 0x100
mask_read_fails_platform = 0x200
mask_read_PCR = 0x400
mask_read_supplementary_alignment = 0x800

bamFileFlagsAnalysis = pysam.AlignmentFile("merged-tumor.bam")

for readed in bamFileFlagsAnalysis:
    counter += 1
    if counter == 1:
        continue
    elif counter > 8:
        break
    else:
        print("Read number " + str(counter) + ", flag value: " + str(readed.flag))

        print("Read Summary:")
        
        if(readed.flag & mask_read_paired):
            print("read paired (0x1)")
        if(readed.flag & mask_read_mapped_in_proper_pair):
            print("read mapped in proper pair (0x2)")
        if(readed.flag & mask_read_unmapped):
            print("read unmapped (0x4)")
        if(readed.flag & mask_mate_unmapped):
            print("mate unmapped (0x8)")
        if(readed.flag & mask_read_reverse_strand):
            print("read reverse strand (0x10)")
        if(readed.flag & mask_mate_reverse_strand):
            print("mate reverse strand (0x20)")
        if(readed.flag & mask_first_in_pair):
            print("first in pair (0x40)")
        if(readed.flag & mask_second_in_pair):
            print("second in pair (0x80)")
        if(readed.flag & mask_not_primary_alignment):
            print("not primary alignment (0x100)")
        if(readed.flag & mask_read_fails_platform):
            print("read fails platform/vendor quality checks (0x200)")
        if(readed.flag & mask_read_PCR):
            print("read is PCR or optical duplicate (0x400)")
        if(readed.flag & mask_read_supplementary_alignment):
            print("supplementary alignment (0x800)")

        print("\n")
bamFileFlagsAnalysis.close()

Read number 2, flag value: 163
Read Summary:
read paired (0x1)
read mapped in proper pair (0x2)
mate reverse strand (0x20)
second in pair (0x80)


Read number 3, flag value: 99
Read Summary:
read paired (0x1)
read mapped in proper pair (0x2)
mate reverse strand (0x20)
first in pair (0x40)


Read number 4, flag value: 99
Read Summary:
read paired (0x1)
read mapped in proper pair (0x2)
mate reverse strand (0x20)
first in pair (0x40)


Read number 5, flag value: 99
Read Summary:
read paired (0x1)
read mapped in proper pair (0x2)
mate reverse strand (0x20)
first in pair (0x40)


Read number 6, flag value: 163
Read Summary:
read paired (0x1)
read mapped in proper pair (0x2)
mate reverse strand (0x20)
second in pair (0x80)


Read number 7, flag value: 99
Read Summary:
read paired (0x1)
read mapped in proper pair (0x2)
mate reverse strand (0x20)
first in pair (0x40)


Read number 8, flag value: 1123
Read Summary:
read paired (0x1)
read mapped in proper pair (0x2)
mate reverse strand (0x20)
fi

<h3>Calculating stats</h3>

In [23]:
bamFileForAnalysis = pysam.AlignmentFile("merged-tumor.bam")

unmappedCounter = 0
mappingZeroQualityCounter = 0
totalNumReads = 0
totalSumQualities = 0
totalSumQualitiesFilteredZeros = 0
countFilteredZeros = 0

for curr in bamFileForAnalysis:
    totalSumQualities += curr.mapping_quality
    totalNumReads += 1
    if(curr.flag & mask_read_unmapped):
        unmappedCounter += 1
    if(curr.mapping_quality == 0):
        mappingZeroQualityCounter += 1
    if(curr.mapping_quality != 0):
        countFilteredZeros += 1
        totalSumQualitiesFilteredZeros += curr.mapping_quality

averageMappingQuality = totalSumQualities / totalNumReads
averageMappingQualityFilteredZeros = totalSumQualitiesFilteredZeros / countFilteredZeros

print("Total number of unmapped reads: ")
print(unmappedCounter)
print("\n")

print("Total number of reads: ")
print(totalNumReads)
print("\n")

print("Number of reads with mapping quality 0: ")
print(mappingZeroQualityCounter)
print("\n")

print("Average mapping quality for all the reads: ")
print(averageMappingQuality)
print("\n")

print("Average mapping quality if reads with 0 map quality are filtered out: ")
print(averageMappingQualityFilteredZeros)
print("\n")

bamFileForAnalysis.close()

Total number of unmapped reads: 
17765


Total number of reads: 
2921629


Number of reads with mapping quality 0: 
126628


Average mapping quality for all the reads: 
55.91379158681681


Average mapping quality if reads with 0 map quality are filtered out: 
58.446975510921106


